File: src/csharp/1_SOLIDDesignPrinciples/3_LiskovSubstitution/LSP.cs
using static System.Console;

namespace DotNetDesignPatternDemos.SOLID.LiskovSubstitutionPrinciple
{
  // using a classic example
  public class Rectangle
  {
    //public int Width { get; set; }
    //public int Height { get; set; }

    public virtual int Width { get; set; }
    public virtual int Height { get; set; }

    public Rectangle()
    {
      
    }

    public Rectangle(int width, int height)
    {
      Width = width;
      Height = height;
    }

    public override string ToString()
    {
      return $"{nameof(Width)}: {Width}, {nameof(Height)}: {Height}";
    }
  }

  public class Square : Rectangle
  {
    //public new int Width
    //{
    //  set { base.Width = base.Height = value; }
    //}

    //public new int Height
    //{ 
    //  set { base.Width = base.Height = value; }
    //}

    public override int Width // nasty side effects
    {
      set { base.Width = base.Height = value; }
    }

    public override int Height
    { 
      set { base.Width = base.Height = value; }
    }
  }

  public class Demo
  {
    static public int Area(Rectangle r) => r.Width * r.Height;

    static void Main(string[] args)
    {
      Rectangle rc = new Rectangle(2,3);
      WriteLine($"{rc} has area {Area(rc)}");

      // should be able to substitute a base type for a subtype
      /*Square*/ Rectangle sq = new Square();
      sq.Width = 4;
      WriteLine($"{sq} has area {Area(sq)}");
    }
  }
}

File: src/swift/1_SOLIDDesignPrinciples/3_LiskovSubstitution/LSP.swift
import Foundation

class Rectangle : CustomStringConvertible
{
  internal var _width: Int = 0
  internal var _height: Int = 0
  var width: Int
  {
    get {
      return _width
    }
    set(value) {
      _width = value
    }
  }
  var height: Int
  {
    get {
      return _height
    }
    set(value) {
      _height = value
    }
  }

  init(_ width: Int, _ height: Int)
  {
    self.width = width
    self.height = height
  }

  public var description: String
  {
    return "Width: \(width), height: \(height)"
  }
}

class Square : Rectangle
{
  var height: Int
  {
    get { return _height }
    set(value)
    {
      _width = value
      _height = value
    }
  }
  var width: Int
  {
    get { return _width }
    set(value)
    {
      _width = value
      _height = value
    }
  }
}

func area(_ r : Rectangle) -> Int {
  return r.width * r.height
}

func main()
{
  var rc = Rectangle(2,3)
  print("\(rc) has area \(area(rc))")

  // should be able to substitute a base type for a subtype
  var sq: Rectangle = Square(0,0)
  sq.width = 4
  print("\(sq) has area \(area(sq))")
}

main()